### PR TITLE
Fix /transaction.all slow responses

### LINKED ISSUE #589: Fix/Research slow /transaction.all endpoint
The endpoint seems to try to load the whole table when querying a list of 200 transactions (see [explorer](http://quest.ari.omg.network/address/0x6f4a4981bc761973eb81e67731e566242c53275a) and end up crashing. I'm thinking about trying to add some indexes.

### PR DESCRIPTION
**Pull request by T-Dnzt**
closes #589

# Overview

This PR switches from using a double left join to using a single inner with a bunch of added indices for increased performance. 

# Note

Also cherry picked [this commit](https://github.com/omisego/elixir-omg/commit/946197ccb372b9be6971abe6587d133e777a63a4) to get it into 0.1 (it was preventing Docker from starting).

### CODE REVIEW

File: apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+defmodule OMG.Watcher.Repo.Migrations.AddMissingIndicesToTxOuputs do
+  use Ecto.Migration
+
+  def change do
+    create index(:txoutputs, [:creating_txhash, :spending_txhash])
+    create index(:txoutputs, [:creating_deposit])
+    create index(:txoutputs, [:spending_txhash])
```
Comment (T-Dnzt) [inline apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs]: This is @kasima – add an NB about why a `spending_txhash` index alone.

File: apps/omg_watcher/lib/db/transaction.ex [hunk 2]
```diff
@@ -21,11 +21,11 @@ defmodule OMG.Watcher.DB.Transaction do
 
   alias OMG.API.State.Transaction
   alias OMG.API.Utxo
-  alias OMG.Watcher.DB
+  alias OMG.Watcher.{DB, DB.TxOutput}
```
Comment (pnowosie) [inline apps/omg_watcher/lib/db/transaction.ex]: I've seen a lot changes actually expanding this kind into separate lines - it's more alchemist ;)

File: apps/omg_watcher/lib/db/transaction.ex [hunk 3]
```diff
@@ -67,7 +67,7 @@ defmodule OMG.Watcher.DB.Transaction do
     DB.Repo.one(query)
   end
 
-  def get_by_filters(address, blknum, limit) do
+  def get_by_filters(address, blknum \\ nil, limit \\ nil) do
```
Comment (pnowosie) [inline apps/omg_watcher/lib/db/transaction.ex]: Please revert - seem to be just for test, and I'm afraid of consequences

File: apps/omg_watcher/lib/db/transaction.ex [hunk 4]
```diff
@@ -83,16 +83,14 @@ defmodule OMG.Watcher.DB.Transaction do
     )
   end
 
-  defp query_get_by_address(base, nil), do: base
+  defp query_get_by_address(query, nil), do: query
```
Comment (pnowosie) [inline apps/omg_watcher/lib/db/transaction.ex]: Nice! Please rename 👇 following code to keep it consistent

File: apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    create index(:txoutputs, [:spending_exit])
```
Comment (pnowosie) [inline apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs]: this ':spending_exit' would be super sparse - consider partial index like [here](https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher/priv/repo/migrations/20190315095855_alter_transactions_table_add_partitial_index.exs)

File: apps/omg_watcher/test/db/transaction_test.exs [hunk 6]
```diff
@@ -24,6 +24,22 @@ defmodule OMG.Watcher.DB.TransactionTest do
 
   require Utxo
 
+  describe "get_by_filters/3" do
```
Comment (pnowosie) [inline apps/omg_watcher/test/db/transaction_test.exs]: Please revert, already tested [here](https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher/test/omg_watcher/web/controllers/transaction_test.exs)

File: apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs (hunk 1 above)
Comment (T-Dnzt) [reply apps/omg_watcher/priv/repo/migrations/20190408131000_add_missing_indices_to_txoutputs.exs]: Composite indexes will be used by the optimizer when either the first column is used, or both columns are used in the query. However, if a query only uses the second column (`spending_txhash`), the composite index will not be used, hence the addition of another index for it.

File: apps/omg_watcher/test/db/transaction_test.exs (hunk 6 above)
Comment (T-Dnzt) [reply apps/omg_watcher/test/db/transaction_test.exs]: But those are unit tests for the `Transaction` schema 🤔Controller tests are not enough to ensure that each function at the schema level works as expected and is thoroughly tested. It also makes sense to build a pyramid, a lot of small unit tests testing a lot of edge cases at the bottom (which I didn't really do because I'm quite unfamiliar with the fixture approach you're using and didn't want to mess up everything), and a few high level integration/controller tests testing the happy path.

File: apps/omg_watcher/lib/db/transaction.ex (hunk 3 above)
Comment (T-Dnzt) [reply apps/omg_watcher/lib/db/transaction.ex]: Removing it, but I do feel having them optional makes the code clearer in terms of expectations (since it isn't handled until 2 functions down in pattern matching).

**Comment by T-Dnzt** [comment]
@pnowosie Merging without removing the tests (because we need the feature kinda right now), but happy to discuss further and potentially submit another PR :)

**Comment by kasima** [comment]
This has  generally improved response times across all of the watcher endpoints. 🙌 

Performance graphs - OmiseGO MoreVP Watcher v20190131 - AppSignal 2019-04-09 14-24-27


## TODO
- Cherry pick this into `master`

File: apps/omg_watcher/test/db/transaction_test.exs (hunk 6 above)
Comment (pdobacz) [reply apps/omg_watcher/test/db/transaction_test.exs]: yeah, we relied on controller tests, since the whole thing was super-simplistic, now as it evolves we can reconsider, in case the controller tests get too unwieldy.
